deep_qa/models/multiple_choice_qa/multiple_choice_tuple_inference.py

```diff
@@ -0,0 +1,167 @@
+            :math:`s^c = 1 - \prod_{i=1:N}(1 - q_2 * s_{i}^c)`
+        where q_2 is the noise parameter for this second noisy-or.  At this point, we have a score for each of
+        the answer candidates, and so we perform a softmax to determine which is the best answer.
+        """
+        # shape: (batch size, num_options, num_question_tuples, num_tuple_slots, num_slot_words)
+        question_input = Input((self.num_options,
+                                self.num_question_tuples,
+                                self.num_tuple_slots,
+                                self.max_slot_length,),
+                               dtype='int32',
+                               name='question_input')
+        # shape: (batch size, num_background_tuples, num_tuple_slots, num_slot_words)
+        background_input = Input((self.num_background_tuples,
+                                  self.num_tuple_slots,
+                                  self.max_slot_length,),
+                                 dtype='int32',
+                                 name='background_input')
+
+        # Expand and tile the question input to be:
+        # shape: (batch size, num_options, num_question_tuples, num_background_tuples, num_tuple_slots,
+        #         num_slot_words)
+        tiled_question = Repeat(axis=3, repetitions=K.int_shape(background_input)[1])(question_input)
+
+        # Expand and tile the background input to match question input.
+        # shape: (batch size, num_options, num_question_tuples, num_background_tuples, num_tuple_slots,
+        #               num_slot_words)
+        # First, add num_options.
+        tiled_background = Repeat(axis=1, repetitions=K.int_shape(question_input)[1])(background_input)
+        # Next, add num_question_tuples.
+        tiled_background = Repeat(axis=2, repetitions=K.int_shape(question_input)[2])(tiled_background)
+
+        # Initialize the layer that does the tuple-matching (i.e., entailment).
+        tuple_match = TupleMatch(**self.entailment_model_params)
+        # These TimeDistributed wrapperss correspond to distributing across each of num_options,
+        # num_question_tuples, and num_background_tuples.
+        match_layer = TimeDistributed(TimeDistributed(TimeDistributed(tuple_match)))
+
+        # Find the matches between the question and background tuples.
+        # shape: (batch size, num_options, num_question_tuples, num_background_tuples, 1)
+        matches = match_layer([tiled_question, tiled_background])
+        # Squeeze to get rid of the last dim of length 1.
+        # shape: (batch size, num_options, num_question_tuples, num_background_tuples)
+        matches = Squeeze(axis=-1)(matches)
+
+        # Find the probability that any given question tuple is entailed by the given background tuples.
+        # shape: (batch size, num_options, num_question_tuples)
+        combine_background_evidence = NoisyOr(axis=-1, param_init=self.noisy_or_param_init, name="noisyOr_1")
```

Comment:
That name is what gets printed out in the Keras model summary.  If you notice, all names there have been converted from camel case (which is appropriate for class names in python) to underscores, or just all lowercase.  So you get this like "vectormatrixmerge", "timedistributed_4" and "word_question_embedding".  This should be either "noisy_or_1", "noisyor_1", or something like "combine_background_evidence".